pkg/auth/auth.go

```diff
@@ -131,32 +153,67 @@ func (u *UserAuth) isAllowed(verb string, itemsToCheck []resource) error {
 		}
 		allowed, _ := u.canPerform(verb, group, resource, i.Namespace)
 		if !allowed {
-			return fmt.Errorf("Unauthorized to %s %s/%s in the %s namespace", verb, i.APIVersion, resource, i.Namespace)
+			rejectedActions = append(rejectedActions, Action{
+				APIVersion: i.APIVersion,
+				Resource:   resource,
+				Namespace:  i.Namespace,
+				Verbs:      []string{verb},
+			})
 		}
 	}
-	return nil
+	return rejectedActions, nil
 }
 
-// CanI returns if the user can perform the given action with the given chart and parameters
-func (u *UserAuth) CanI(namespace, action, manifest string) error {
+func reduceActionsByVerb(actions []Action) []Action {
+	resMap := map[string]Action{}
+	res := []Action{}
+	for _, action := range actions {
+		req := fmt.Sprintf("%s/%s/%s", action.Namespace, action.APIVersion, action.Resource)
+		if _, ok := resMap[req]; ok {
+			// Element already exists
+			verbs := append(resMap[req].Verbs, action.Verbs...)
+			resMap[req] = Action{
+				APIVersion: action.APIVersion,
+				Resource:   action.Resource,
+				Namespace:  action.Namespace,
+				Verbs:      verbs,
+			}
+		} else {
+			resMap[req] = action
+		}
+	}
+	for _, a := range resMap {
+		res = append(res, a)
+	}
+	return res
+}
+
+// CanI checks if the current user can perform the given action on all the resources given in the manifests
+// It returns the list of Actions that the user cannot perform
```

Comment:
Can I suggest creating a custom error struct that includes the list of forbidden/rejected actions? IMO this function should return a true/false to indicate whether or not the action can be performed.

Reply:
It is true that `CanI` infers that it should return `true|false` but in what I am really interested is to return the list of forbidden actions (and in case there is an unexpected error return it). It can also be confused with the `can-i` command of `kubectl` while this is slightly different. I will rename the method and rephrase the docs.